codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,457 @@
+        }
+
+        @Override
+        public void onStreamClosed(Http2Stream stream) {
+            final StreamInfo streamInfo = (StreamInfo) stream.removeProperty(streamInfoKey);
+            if (streamInfo != null) {
+                if (streamInfo.childChannel.eventLoop().inEventLoop()) {
+                    onStreamClosed0(streamInfo);
+                } else {
+                    streamInfo.childChannel.eventLoop().execute(new OneTimeTask() {
+                        @Override
+                        public void run() {
+                            onStreamClosed0(streamInfo);
+                        }
+                    });
+                }
+            }
+        }
+
+        private void onStreamClosed0(StreamInfo streamInfo) {
+            streamInfo.childChannel.onStreamClosedFired = true;
+            streamInfo.childChannel.fireChildRead(new AbstractHttp2StreamChannel.CloseMessage());
+        }
+
+        @Override
+        public void onGoAwayReceived(final int lastStreamId, long errorCode, ByteBuf debugData) {
+            final Http2GoAwayFrame goAway = new DefaultHttp2GoAwayFrame(errorCode, debugData);
+            try {
+                http2Handler.connection().forEachActiveStream(new Http2StreamVisitor() {
+                    @Override
+                    public boolean visit(Http2Stream stream) {
+                        if (stream.id() > lastStreamId
+                                && http2Handler.connection().local().isValidStreamId(stream.id())) {
+                            StreamInfo streamInfo = (StreamInfo) stream.getProperty(streamInfoKey);
+                            if (streamInfo != null) {
+                                streamInfo.childChannel.pipeline()
+                                    .fireUserEventTriggered(goAway.retain());
+                            }
+                        }
+                        return true;
+                    }
+                });
+            } catch (Throwable t) {
+                ctx.invoker().invokeExceptionCaught(ctx, t);
+            }
+            ctx.fireUserEventTriggered(goAway.retain());
+        }
+    }
+
+    class InternalHttp2ConnectionHandler extends Http2ConnectionHandler {
+        public InternalHttp2ConnectionHandler(Http2Connection connection,
+                Http2FrameReader frameReader, Http2FrameWriter frameWriter,
+                Http2FrameListener listener) {
+            super(connection, frameReader, frameWriter, listener);
+        }
+
+        @Override
+        protected void onStreamError(ChannelHandlerContext ctx, Throwable cause,
+                Http2Exception.StreamException http2Ex) {
+            try {
+                Http2Stream stream = http2Handler.connection().stream(http2Ex.streamId());
+                if (stream == null) {
+                    return;
+                }
+                StreamInfo streamInfo = (StreamInfo) stream.getProperty(streamInfoKey);
+                if (streamInfo == null) {
```

Comment:
This is the only interesting case left. Because it is handling an error, we wouldn't want to cause a cascade of errors. If I've removed the check due to GO_AWAY, is this one still confusing and deserves a comment?